packages/flutter/lib/src/services/text_input.dart

```diff
@@ -2015,3 +2099,257 @@ class TextInput {
+  ///
+  /// This method is called when a text input client is attached. The input
+  /// control should update its configuration to match the client's configuration.
+  void attach(TextInputClient client, TextInputConfiguration configuration) {}
+
+  /// Requests the text input control to detach from the given input client.
+  ///
+  /// This method is called when a text input client is detached. The input
+  /// control should release any resources allocated for the client.
+  void detach(TextInputClient client) {}
+
+  /// Requests that the text input control is shown.
+  ///
+  /// This method is called when the input control should become visible.
+  void show() {}
+
+  /// Requests that the text input control is hidden.
+  ///
+  /// This method is called when the input control should hide.
+  void hide() {}
+
+  /// Informs the text input control about input configuration changes.
+  ///
+  /// This method is called when the configuration of the attached input client
+  /// has changed.
+  void updateConfig(TextInputConfiguration configuration) {}
+
+  /// Informs the text input control about editing state changes.
+  ///
+  /// This method is called when the editing state of the attached input client
+  /// has changed.
+  void setEditingState(TextEditingValue value) {}
+
+  /// Informs the text input control about client position changes.
+  ///
+  /// This method is called on when the input control should position itself in
+  /// relation to the attached input client.
+  void setEditableSizeAndTransform(Size editableBoxSize, Matrix4 transform) {}
+
+  /// Informs the text input control about composing area changes.
+  ///
+  /// This method is called when the attached input client's composing area
+  /// changes.
+  void setComposingRect(Rect rect) {}
+
+  /// Informs the text input control about caret area changes.
+  ///
+  /// This method is called when the attached input client's caret area
+  /// changes.
+  void setCaretRect(Rect rect) {}
+
+  /// Informs the text input control about selection area changes.
+  ///
+  /// This method is called when the attached input client's selection area
+  /// changes.
+  void setSelectionRects(List<SelectionRect> selectionRects) {}
+
+  /// Informs the text input control about text style changes.
+  ///
+  /// This method is called on the when the attached input client's text style
+  /// changes.
+  void setStyle({
+    required String? fontFamily,
+    required double? fontSize,
+    required FontWeight? fontWeight,
+    required TextDirection textDirection,
+    required TextAlign textAlign,
+  }) {}
+
+  /// Requests autofill from the text input control.
+  ///
+  /// This method is called when the autofill UI should appear.
+  void requestAutofill() {}
+
+  /// Requests that the autofill context is finalized.
+  ///
+  /// See also:
+  ///
+  ///  * [TextInput.finishAutofillContext]
+  void finishAutofillContext({bool shouldSave = true}) {}
```

Comment:
Similar question as above about these empty implementations.  Classes that extend TextInputControl will not be required to implement all of the methods here, is that ok? (Classes that use `implements` will have to implement them all regardless of if there is an empty implementation here or not.)

Reply:
It's ok to not implement all `TextInputControl`. A pure abstract interface that I started with was inconvenient to work with because many examples didn't care about half of the methods.

I think it was a great idea to make `TextInputClient` and friends mixins. I changed `TextInputControl` to a mixin too.

Reply:
Ok great, then empty implementations sound good to me 👍